src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
+Verona also doesn't restrict declaration order. Functions and classes can be declared at any time in the module to be used anywhere.
+To avoid order issues in the IR, we'll do multiple passes on the structures:
+1. Declaration pass: Defines reachable type structures and functions, including the entire standard library files.
+2. Lowering pass: Lower the reachable AST nodes using the structures above and defining the functions for the prototypes above.
+
+In theory, the AST could have been directly lowered to LLVM IR, but the LLVM IR is too low level for some analysis we want to do (ex. region aliasing) and it would be counter-productive to do so at that stage.
+
+## Classes & Modules
+
+Classes and modules are the same concept in Verona.
+Each source sub-directory containing Verona files is a module, and all its Verona files share the same namespace, named after the directory name.
+The base source directory is also a module, and its sub-directories will be included in the "path" for importing modules.
+
+Calling for a method or accessing a field on an object of a type in a different module needs full scope.
+The parser will resolve all those issues and will cannonicalise the representation for all access to be fully scoped.
+
+Class types will be represented in a mix of LLVM structure types (for fields) and a `vtable` (as an array of pointers) for the dynamic methods.
+
+## Functions
+
+Functions are a native concept in MLIR and LLVM IR, and lowering free and static functions and calls to them is trivial.
+We'll declare all reachable functions beforehand, then lower the AST (implementing the prototypes as full functions).
+Dynamic resolution and selector colouring will be done at the parser level, and calls to dynamic methods in the AST will be a node with base pointer and offset.
+
+MLIR modules have a symbol table for functions, which we use to match declarations against definitions and calls.
+
+Arguments by value or reference have ABI reserved space (registers, stack), so we don't need to allocate them again (the `alloca` _trick_).
+Return values, for functions that do return a type, will be the last expression evaluated in the block.
+
+Lambdas will be lowered as an anonymous structure with captures as fields and `apply()` as the method for when the lambda is called.
+That transformation should have already been done by the parser and represented in the AST by the right basic constructs.
+At the MLIR level, they'll be allocated and dispatched in the usual way (stack/heap + call/behaviour dispatch).
+
+**QUESTION: How do we mark the `main` function?**
+Currently, if there is a function called `main`, it's lowered without scope to help testing execution, but we need more strict rules, or perhaps annotation, to mark the entry point.
+We could also have a class named `Main` which is restricted to only one per program, with a method `main(int, int**) : int`, which is the entry point.
```

Comment:
We could also adopt the OpenStep approach.  I'm somewhat biased here, but it's the one that I prefer.  Each program build has a property list that defines a bunch of metadata that doesn't live in any source files (author, version number, and so on) and this includes the name of the application class.  The entry point in the program is in this class.  This makes it fairly easy to build multiple programs with different entry points (e.g. for testing).

Reply:
I like the idea of annotating the entry point. The implementation is simpler and clearer. But this needs to come from the AST, which in turn, comes from the language syntax and user code.

I hate what Python does with the `__main__` thing. I hate even more what Perl does, starting from line 1.

I'd prefer either a static agreed name (like Java's `class Main` + `function main(...)`) or an annotation on the function, like:
```
func blergh(argc: I32, argv: [CharString]) : I32 # entry_point
{
  ...
}
```

Reply:
To be clear: I want to keep the question here for this PR. I don't mean to answer it right now, as this is a complex subject that needs discussion.